csharp/ql/lib/semmle/code/csharp/security/auth/MissingFunctionLevelAccessControlQuery.qll

```diff
@@ -0,0 +1,184 @@
+}
+
+/** An action method in the MVC framework. */
+private class MvcActionMethod extends ActionMethod {
+  MvcActionMethod() { this = any(MicrosoftAspNetCoreMvcController c).getAnActionMethod() }
+}
+
+/** An action method on a subclass of `System.Web.UI.Page`. */
+private class WebFormActionMethod extends ActionMethod {
+  WebFormActionMethod() {
+    this.getDeclaringType().getBaseClass*() instanceof SystemWebUIPageClass and
+    this.getAParameter().getType().getName().matches("%EventArgs")
+  }
+
+  override Callable getAnAuthorizingCallable() {
+    result = this
+    or
+    result.getDeclaringType() = this.getDeclaringType() and
+    result.getName() = "Page_Load"
+  }
+
+  override string getARoute() {
+    exists(string physicalRoute | physicalRoute = super.getARoute() |
+      result = physicalRoute
+      or
+      exists(string absolutePhysical |
+        virtualRouteMapping(result, absolutePhysical) and
+        physicalRouteMatches(absolutePhysical, physicalRoute)
+      )
+    )
+  }
+}
+
+/**
+ * Holds if `virtualRoute` is a URL path
+ * that can map to the corresponding `physicalRoute` filepath
+ * through a call to `MapPageRoute`
+ */
+private predicate virtualRouteMapping(string virtualRoute, string physicalRoute) {
+  exists(MethodCall mapPageRouteCall, StringLiteral virtualLit, StringLiteral physicalLit |
+    mapPageRouteCall
+        .getTarget()
+        .hasQualifiedName("System.Web.Routing", "RouteCollection", "MapPageRoute") and
+    virtualLit = mapPageRouteCall.getArgument(1) and
+    physicalLit = mapPageRouteCall.getArgument(2) and
+    virtualLit.getValue() = virtualRoute and
+    physicalLit.getValue() = physicalRoute
+    // physicalRouteMatches(physicalLit.getValue(), physicalRoute)
+  )
+}
+
+/** Holds if the filepath `route` can refer to `actual` after expanding a '~". */
+bindingset[route, actual]
+private predicate physicalRouteMatches(string route, string actual) {
+  route = actual
+  or
+  route.charAt(0) = "~" and
+  exists(string dir | actual = dir + route.substring(1, route.length()) + ".cs")
+}
+
+/** An expression that indicates that some authorization/authentication check is being performed. */
+class AuthExpr extends Expr {
+  AuthExpr() {
+    this.(MethodCall)
+        .getTarget()
+        .hasQualifiedName("System.Security.Principal", "IPrincipal", "IsInRole")
+    or
+    this.(PropertyAccess)
+        .getTarget()
+        .hasQualifiedName("System.Security.Principal", "IIdentity", ["IsAuthenticated", "Name"])
+    or
+    this.(MethodCall).getTarget().getName().toLowerCase().matches("%auth%")
+    or
+    this.(PropertyAccess).getTarget().getName().toLowerCase().matches("%auth%")
+  }
+}
+
+/** Holds if `m` is a method that should have an auth check, and does indeed have one. */
+predicate hasAuthViaCode(ActionMethod m) {
+  m.needsAuth() and
+  exists(Callable caller, AuthExpr auth |
+    m.getAnAuthorizingCallable().calls*(caller) and
```

Comment:
Could this cause performance issues (the reflexive and transitive closure of this could be huge)?

Reply:
Alright - that sounds good!

Reply:
@hvitved: Do you have any concerns about using the reflexive, transitive closure of `calls`?

Reply:
If DCA is happy, then I'm happy :-)